grl/policy_eval.py

```diff
@@ -87,20 +102,32 @@ def get_weights(self, no_gamma):
         b = -1 * self.amdp.p0 # subtract p0(s) to left side
         return np.linalg.solve(a, b)
 
-    def solve_amdp(self, mdp_vals, weights):
+    def solve_amdp(self, mdp_vals, occupancy):
         """
         Weights the value contribution of each state to each observation for the amdp
         """
-        amdp_vals = np.zeros(self.amdp.n_obs)
-        for i in range(self.amdp.n_obs):
-            col = self.amdp.phi[:,i].copy().astype('float')
-            col *= weights
+        v_vals = np.zeros(self.amdp.n_obs)
+        q_vals = np.zeros((self.amdp.n_obs, self.amdp.n_actions))
+        for o in range(self.amdp.n_obs):
+            # V vals
+            col = self.amdp.phi[:,o].copy().astype('float')
+            col *= occupancy
             col /= col.sum()
             v = mdp_vals * col
-            # amdp_vals[i] += v.sum()
-            amdp_vals = amdp_vals.at[i].set(amdp_vals[i] + v.sum())
+            # v_vals[o] += v.sum()
+            v_vals = v_vals.at[o].set(v_vals[o] + v.sum())
+
+            # Q vals
+            for s in range(self.amdp.n_states):
+                for a in range(self.amdp.n_actions):
+                    for sp in range(self.amdp.n_states):
+                        q_vals = q_vals.at[
+                            o, a].set(q_vals[o, a] + col[s] *
+                                      (self.amdp.T[a, s, sp] *
+                                       (self.amdp.R[a, s, sp] +
+                                        self.amdp.gamma * mdp_vals[sp])))
```

Comment:
I think we should send in the `mdp_qvals`, use them to calculate `amdp_q_vals`, and then compute `amdp_v_vals` from there. We can save ourselves from this nasty triple-for-loop

Reply:
Will look into this

Reply:
Nice, done.